okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt

```diff
@@ -0,0 +1,921 @@
+    streamId: Int,
+    unacknowledgedBytesRead: Long
+  ) {
+    writerExecutor.tryExecute("OkHttp Window Update $connectionName stream $streamId") {
+      try {
+        writer.windowUpdate(streamId, unacknowledgedBytesRead)
+      } catch (e: IOException) {
+        failConnection(e)
+      }
+    }
+  }
+
+  fun writePing(
+    reply: Boolean,
+    payload1: Int,
+    payload2: Int
+  ) {
+    if (!reply) {
+      val failedDueToMissingPong: Boolean
+      synchronized(this) {
+        failedDueToMissingPong = awaitingPong
+        awaitingPong = true
+      }
+      if (failedDueToMissingPong) {
+        failConnection(null)
+        return
+      }
+    }
+
+    try {
+      writer.ping(reply, payload1, payload2)
+    } catch (e: IOException) {
+      failConnection(e)
+    }
+  }
+
+  /** For testing: sends a ping and waits for a pong.  */
+  @Throws(InterruptedException::class)
+  fun writePingAndAwaitPong() {
+    writePing(false, 0x4f4b6f6b /* "OKok" */, -0xf607257 /* donut */)
+    awaitPong()
+  }
+
+  /** For testing: waits until `requiredPongCount` pings have been received from the peer.  */
+  @Synchronized @Throws(InterruptedException::class)
+  fun awaitPong() {
+    while (awaitingPong) {
+      (this as Object).wait()
+    }
+  }
+
+  @Throws(IOException::class)
+  fun flush() {
+    writer.flush()
+  }
+
+  /**
+   * Degrades this connection such that new streams can neither be created locally, nor accepted
+   * from the remote peer. Existing streams are not impacted. This is intended to permit an endpoint
+   * to gracefully stop accepting new requests without harming previously established streams.
+   */
+  @Throws(IOException::class)
+  fun shutdown(statusCode: ErrorCode) {
+    synchronized(writer) {
+      val lastGoodStreamId: Int
+      synchronized(this) {
+        if (isShutdown) {
+          return
+        }
+        isShutdown = true
+        lastGoodStreamId = this.lastGoodStreamId
+      }
+      // TODO: propagate exception message into debugData.
+      // TODO: configure a timeout on the reader so that it doesn’t block forever.
+      writer.goAway(lastGoodStreamId, statusCode, Util.EMPTY_BYTE_ARRAY)
+    }
+  }
+
+  /**
+   * Closes this connection. This cancels all open streams and unanswered pings. It closes the
+   * underlying input and output streams and shuts down internal executor services.
+   */
+  override fun close() {
+    close(ErrorCode.NO_ERROR, ErrorCode.CANCEL, null)
+  }
+
+  internal fun close(
+    connectionCode: ErrorCode,
+    streamCode: ErrorCode,
+    cause: IOException?
+  ) {
+    assert(!Thread.holdsLock(this))
+    ignoreIoExceptions {
+      shutdown(connectionCode)
+    }
+
+    var streamsToClose: Array<Http2Stream>? = null
+    synchronized(this) {
+      if (!streams.isEmpty()) {
+        streamsToClose = streams.values.toTypedArray()
+        streams.clear()
+      }
+    }
+
+    if (streamsToClose != null) {
+      for (stream in streamsToClose!!) {
```

Comment:
ugh. could be `streamsToClose?.forEach { stream -> .. }` if you want

Reply:
Done.

Reply:
Shouldn’t be a problem with contracts?